### DISCUSSION TITLE
TestStore and snapshot testing

### DISCUSSION THREAD

**Discussion by dmeehan1968** [pointfreeco/swift-composable-architecture#885] [Q&A] [answered]
Looking at the example code in [Episode 86](https://www.pointfree.co/episodes/ep86-swiftui-snapshot-testing), there is a simplified implementation of TCA, via the ComposableArchitecture.swift module.  It's not clear why this is necessary, but the API is not the same as the full TCA framework.

This alternative implementation seems to provide a facility similar to TestStore, in that it allows for actions to be rooted to the store.  The problem is that the type of the TestStore isn't compatible with the store being passed to the view, and there doesn't appear to be a method to produce one that is compatible (`scope` produces a similar incompatible type).

That seems to limit snapshot testing to setting up store state for each intended snapshot, without the ability to then send actions prior to another snapshot.

If its just a case of needing to use the simplified implementation in the episode 86 example, perhaps that can be clarified.

**Comment by dmeehan1968** [comment, marked as answer]
This was answered here - https://github.com/pointfreeco/swift-composable-architecture/discussions/879#discussioncomment-1614317

**Reply by remlostime** [reply]
hmm, the answer seems not clear to me.  @dmeehan1968 Do you mind to explain it a little bit?
> For snapshot testing, since in View, we use ViewStore to send the action, and we cannot access ViewStore in our snapshot test. How do we mimic each action flow to the ViewStore and compare the screenshot?

**Reply by dmeehan1968** [reply]
In simple terms, TestStore isn't designed for snapshot testing, it's designed to simplify testing of reducer logic.

If your views are correctly a function of state, then you can create snapshots with any representation of state, via a full store.  If your state is particularly complex you could use TestStore to drive state changes, then take the resulting state and pass this to a new full store as initialState, then use that to generate the snapshot. 

It's worth bearing in mind that you've coupled the reducer and view representation in doing this, which potentially makes your tests more brittle (you don't need to test a snapshot if you know the state is incorrect).
